pkg/storage/etcd/etcd_watcher_test.go

```diff
@@ -320,405 +283,163 @@ func makeSubsets(ip string, port int) []api.EndpointSubset {
+	if event.Type != watch.Added {
+		t.Errorf("Unexpected event %#v", event)
+	}
 
-		t.Logf("Testing %v", k)
-		for i := range testCase.Responses {
-			if testCase.Responses[i] != nil {
-				fakeClient.WatchResponse <- testCase.Responses[i]
-			}
-			event := <-watching.ResultChan()
-			if e, a := testCase.Expected[i].Type, event.Type; e != a {
-				t.Errorf("%s: expected type %v, got %v", k, e, a)
-				break
-			}
-			if e, a := testCase.Expected[i].Endpoints, event.Object.(*api.Endpoints).Subsets; !api.Semantic.DeepDerivative(e, a) {
-				t.Errorf("%s: expected type %v, got %v", k, e, a)
-				break
-			}
-		}
-		watching.Stop()
+	subset := makeSubsets("127.0.0.1", 9000)
+	endpoint.Subsets = subset
+
+	// CAS the previous value
+	err = h.Set(context.TODO(), key, endpoint, endpoint, 0)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	event = <-watching.ResultChan()
+	if event.Type != watch.Modified {
+		t.Errorf("Unexpected event %#v", event)
 	}
+
+	if e, a := endpoint, event.Object; !api.Semantic.DeepDerivative(e, a) {
+		t.Errorf("%s: expected %v, got %v", e, a)
+	}
+
+	watching.Stop()
 }
 
 func TestWatchFromZeroIndex(t *testing.T) {
 	codec := testapi.Default.Codec()
 	pod := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}}
 
-	testCases := map[string]struct {
-		Response        tools.EtcdResponseWithError
-		ExpectedVersion string
-		ExpectedType    watch.EventType
-	}{
-		"get value created": {
-			tools.EtcdResponseWithError{
-				R: &etcd.Response{
-					Node: &etcd.Node{
-						Value:         runtime.EncodeOrDie(codec, pod),
-						CreatedIndex:  1,
-						ModifiedIndex: 1,
-					},
-					Action:    "get",
-					EtcdIndex: 2,
-				},
-			},
-			"1",
-			watch.Added,
-		},
-		"get value modified": {
-			tools.EtcdResponseWithError{
-				R: &etcd.Response{
-					Node: &etcd.Node{
-						Value:         runtime.EncodeOrDie(codec, pod),
-						CreatedIndex:  1,
-						ModifiedIndex: 2,
-					},
-					Action:    "get",
-					EtcdIndex: 3,
-				},
-			},
-			"2",
-			watch.Modified,
-		},
-	}
-
-	for k, testCase := range testCases {
-		fakeClient := tools.NewFakeEtcdClient(t)
-		key := "/some/key"
-		prefixedKey := etcdtest.AddPrefix(key)
-		fakeClient.Data[prefixedKey] = testCase.Response
-		h := newEtcdHelper(fakeClient, codec, etcdtest.PathPrefix())
+	key := etcdtest.AddPrefix("/somekey/foo")
+	server := NewEtcdTestClientServer(t)
+	defer server.Terminate(t)
 
-		watching, err := h.Watch(context.TODO(), key, 0, storage.Everything)
-		if err != nil {
-			t.Fatalf("Unexpected error: %v", err)
-		}
+	h := newEtcdHelper(server.client, codec, etcdtest.PathPrefix())
 
-		fakeClient.WaitForWatchCompletion()
-		if e, a := testCase.Response.R.EtcdIndex+1, fakeClient.WatchIndex; e != a {
-			t.Errorf("%s: expected watch index to be %d, got %d", k, e, a)
-		}
+	// set before the watch and verify events
+	err := h.Set(context.TODO(), key, pod, pod, 0)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
```

Comment:
Can you please add CAS operation here too? I would like to verify two operations touching the same object that happened before are concatenated (I believe they should be, right?)